### ISSUE TITLE
No meaningful error message when cipher-list do not match

### ISSUE THREAD

**Issue by wisotzky** [grpc/grpc#10702]
While I am building a gNMI telemetry receiver, I was really struggling to figure out, that a 3rd party gRPC server and my gRPC client did not use the same set of ciphers. In the specific case, both did not came together (no matches) - but there was no good error message indicating this.

I would expect to receive an exception (in Python) to indicate, that "no matching cipher" have been found and connection could not be established for that reason. When enabling log-level to debug, I would even expect to see, what is contained in the local/remote cipher list to adapt changes.

Any way, to enable this kind of debug? Are there specific debugs to enable on SSL/TLS level, or just enough to turn GRPC logging to its maximum. If there is something on SSL/TLS - how is this done?

**Comment by jboeuf** [comment]
as @nathanielmanistaatgoogle said in #10701.
Setting GRPC_TRACE and GRPC_VERBOSITY should certainly work.
The trace for tls/ssl in particular is called "transport_security".

**Comment by wisotzky** [comment]
I am now using the following settings:
```
# export GRPC_TRACE=transport_security
# export GRPC_VERBOSITY=debug
```

Those are the debugs I am receiving:
```
I0427 08:35:54.732603695   16024 ssl_transport_security.c:193]      HANDSHAKE START -  before connect initialization  - CINIT
I0427 08:35:54.732662083   16024 ssl_transport_security.c:193]                 LOOP -  before connect initialization  - CINIT
I0427 08:35:54.732748500   16024 ssl_transport_security.c:193]                 LOOP -     SSLv3 write client hello A  - 3WCH_A
I0427 08:35:54.732759659   16024 ssl_transport_security.c:193]                 LOOP -               SSLv3 flush data  - 3FLUSH
D0427 08:35:54.735555581   16024 security_handshaker.c:126]  Security handshake failed: {"created":"@1493278554.735513685","description":"Handshake read failed","file":"src/core/lib/security/transport/security_handshaker.c","file_line":238,"referenced_errors":[{"created":"@1493278554.735507582","description":"Socket closed","fd":6,"file":"src/core/lib/iomgr/tcp_posix.c","file_line":250,"target_address":"ipv4:172.20.148.52:57400"}]}
I0427 08:35:54.735672540   16024 subchannel.c:705]           Connect failed: {"created":"@1493278554.735513685","description":"Handshake read failed","file":"src/core/lib/security/transport/security_handshaker.c","file_line":238,"referenced_errors":[{"created":"@1493278554.735507582","description":"Socket closed","fd":6,"file":"src/core/lib/iomgr/tcp_posix.c","file_line":250,"target_address":"ipv4:172.20.148.52:57400"}]}
I0427 08:35:54.735696835   16024 subchannel.c:503]           Retry in 19.996015760 seconds
17/04/27 08:35:54,736 ERROR    MainThread grpc.RpcError received:
<bound method _Rendezvous.details of <_Rendezvous of RPC that terminated with (StatusCode.UNAVAILABLE, Connect Failed)>>
```

These are the logs from the client implementation. What I was hoping to see is the cipher-suites offered from the server and the cipher-suites supported by the client. Also I would like to see, which side (client or server) is turning down the connection and why.

In the specific example the gRPC/TLS server enforces TLS security. The certificate is validated with success against the root certificate and provides the IP@ of the server as subjectAltName. But cipher suites does not match, while the server does not allow any unencrypted  transmission to be used as fallback. So I believe the logs are correct, as the socket is just closed by the server when it finds out, that it can not establish an encrypted TLS connectivity - but I would like to see in the logs (trace level is debug) that the certificate is good and what ciphers are exchanged.

I don't think this limited logs help to understand, what is going wrong!

**Comment by wisotzky** [comment]
Potentially, once you've got a more complete trace, you can see in which phase the TLS session setup the thing breaks. For this it is important to know, that first thing happens is cipher exchange. This is even before the server certificate will be send. If you see the TCP connection is established but stops after the Client Hello (including flush) - this is clear indication that the server does not like something in the hello which is most likely the ciphers.

The bad reality with my Python client is, that I am always receiving the same grpc.RpcError:
```
<bound method _Rendezvous.details of <_Rendezvous of RPC that terminated with (StatusCode.UNAVAILABLE, Connect Failed)>>
```

This is for all of the following potential failure scenarios:
- gRPC server not started (not listening on TCP port)
- Running gRPC client against wrong server port 21/22/23 (TELNET, SSHv2, FTP port)
- None of the cipher suites provided is supported by server (server closes connection)
- Server certificate validation failed (check against CA root certificate failed)
- Server certificate validation failed (subjectAltName / CN do not fit)

I would be more than happy, if the grpc.RpcError would contain information enough to identify the issue and troubleshoot the issue immediately. Enabling gRPC debug-level on transport_security indeed provide some indication about the issue - but clearly this is not good enough for production use.

**Comment by nathanielmanistaatgoogle** [comment]
@dgquintas and @ctiller: what are the chances that Core would ever emit status details strings more descriptive than "Connect Failed" for any or all of [those scenarios](https://github.com/grpc/grpc/issues/10702#issuecomment-297710996)? Or should @wisotzky be looking elsewhere for more information, like in the metadata?

**Comment by wisotzky** [comment]
Any update?

**Comment by MaxFangX** [comment]
@wisotzky I ran into a similar issue as well where I needed to see the cipher suite messages. The key here is that gRPC is trying to use SSLv3, which is old and insecure.

Running `export GRPC_SSL_CIPHER_SUITES=ECDHE-ECDSA-AES256-GCM-SHA384` to force the client to use a a different cipher suite worked for me. 

Solution found [here](https://groups.google.com/forum/#!topic/grpc-io/VhJAONcDI3c)

**Comment by ncteisen** [comment]
Sounds like the root issue is resolved.

The feature brought up in the comments, tracked in #13790, is being implemented in #13689
